feat: migrations generate cmd

alvarosabu opened the pull request:
# Migrations Generate Command

The `storyblok migrations generate` command allows you to generate migration files for specific component fields. This is useful when you need to transform or update field values across your Storyblok content.

## Basic Usage

```bash
# Generate a migration for a specific component 
storyblok migrations generate my-component --space 295017

# Generate a migration with a custom path
storyblok migrations generate my-component --space 295017 --path ./custom-path

# Generate a migration with a custom suffix
storyblok migrations generate my-component --space 295017 --suffix fieldname
```

## Command Options

| Option | Alias | Description |
|--------|-------|-------------|
| `--space <spaceId>` | `-s` | (Required) The ID of the space to generate migrations for |
| `--path <path>` | `-p` | Custom path to store the migration files (default: ".storyblok/migrations/{spaceId}") |
| `--suffix <suffix>` | `--su` | Suffix to add to the file name (e.g. {component-name}-{field}.<suffix>.js) |
| `--verbose` | `-v` | Show detailed logs and error messages |

## Output Structure

The command will create a migration file in the following structure:

```markdown
.storyblok/
└── migrations/
    └── {spaceId}/
        └── {component-name}.js
```

If a suffix is provided, the file will be named:
```markdown
.storyblok/
└── migrations/
    └── {spaceId}/
        └── {component-name}.<suffix>.js
```

## Migration File Structure

Each generated migration file follows this structure:

```javascript
export default function (block) {
  // Example to change a string to boolean
  // block.fullname = !!(block.fullname)

  // Example to transfer content from other field
  // block.fullname = block.other_field
}
```

## Examples

### Generate a Basic Migration

```bash
storyblok migrations generate hero --space 295017
```

This will create:

```markdown
.storyblok/
└── migrations/
    └── 295017/
        └── hero.js
```

### Generate with Custom Path

```bash
storyblok migrations generate hero --space 295017 --path ./migrations
```

This will create:

```markdown
migrations/
└── 295017/
    └── hero.js
```

### Generate with Custom Suffix

```bash
storyblok migrations generate hero --space 295017 --suffix staging
```

This will create:

```markdown
.storyblok/
└── migrations/
    └── 295017/
        └── hero.staging.js
```

## Error Cases

The command will fail with helpful error messages in the following cases:

- No component name provided
- No field name provided
- No space ID provided
- User not logged in
- Component not found in the space
- Invalid space ID

## Manual Testing Checklist

1. Basic Generation
   - [x] Generate migration for a simple component
   - [x] Verify file is created in the correct location
   - [x] Check if generated file has correct structure
   - [x] Verify component and field names are correctly set

2. Path Options
   - [x] Test default path generation
   - [x] Test custom path generation
   - [x] Test custom suffix generation
   - [x] Verify directory structure is maintained

3. Error Handling
   - [ ] Test without component name
   - [x] Test without space ID
   - [ ] Test with non-existent component
   - [ ] Test while not logged in
   - [ ] Verify error messages are clear and helpful

4. Special Cases
   - [x] Test with complex component names
   - [ ] Test with nested field paths
   - [x] Test verbose mode output
   - [x] Test with various suffix values
   - [x] Verify file permissions are correct

## Best Practices

1. **Naming Conventions**
   - Use clear, descriptive names for your migration files
   - Follow a consistent pattern for field names
   - Use meaningful suffixes (e.g., 'staging', 'prod', 'v1')

2. **Migration Logic**
   - Keep migrations atomic (one change per migration)
   - Always include validation in your migration logic
   - Test migrations on a staging space first

3. **Version Control**
   - Commit migration files to version control
   - Include clear documentation of what each migration does
   - Use suffixes to differentiate between environments

## Related Commands

- `storyblok migrations run` - Apply generated migrations to your content

alvarosabu committed:
feat: add migrations command with generate functionality

- Introduced new migrations command with generate subcommand
- Added constants for migrations command and color palette
- Implemented migration file generation for specific component fields
- Created actions and tests for migration file creation
- Updated index to import migrations command and related modules

alvarosabu committed:
Merge branch 'feature/push-components-cmd' into feature/migrations-generate-cmd

alvarosabu committed:
refactor: remove debug logging and optimize import order in push and migrations commands

alvarosabu committed:
test: remove debug logging from push command test

alvarosabu committed:
test: enhance migrations generate command test coverage

- Added comprehensive test cases for migrations generate command
- Implemented tests for successful migration generation with and without custom path
- Added error handling tests for missing component name, field, and non-existent components
- Updated test file to improve error checking and mock session state

alvarosabu committed:
docs: add comprehensive README for migrations generate command

- Created detailed documentation for migrations generate command
- Included usage examples, command options, and output structure
- Added manual testing checklist and best practices
- Provided insights into migration file generation and usage scenarios

alvarosabu committed:
feat: add optional suffix for migration file generation

- Enhanced migration generation to support optional file name suffix
- Updated `generateMigration` function to accept optional suffix parameter
- Modified command options to include `--suffix` flag
- Updated README with new suffix generation example and documentation
- Adjusted file name generation logic to incorporate optional suffix

alvarosabu committed:
test: update migration generation tests to support optional suffix parameter

- Modified test cases to include optional suffix parameter in `generateMigration` function calls
- Ensured test coverage reflects the recent enhancement of migration file generation

alvarosabu committed:
Merge branch 'next' into feature/migrations-generate-cmd

alvarosabu committed:
refactor: removed field option

- Removed required field parameter from migration generation
- Updated migration template with more generic examples
- Simplified file naming convention for migration files
- Removed unnecessary field-specific logic from generate command
- Updated tests to reflect new migration generation approach

alexjoverm commented on src/commands/migrations/generate/README.md:
````diff
@@ -0,0 +1,166 @@
+
+```bash
+# Generate a migration for a specific component field
+storyblok migrations generate my-component --field my-field --space 295017
````
@alvarosabu shouldn't the --field flag be a more specific case rather than the common one? IMO, we can just have 1 of the examples with the --field flag and the rest without it

edodusi replied:
actually `--field` is not an option of the generate command

alexjoverm commented on src/commands/migrations/generate/README.md:
```diff
@@ -0,0 +1,166 @@
+| Option | Alias | Description |
+|--------|-------|-------------|
+| `--space <spaceId>` | `-s` | (Required) The ID of the space to generate migrations for |
+| `--field <fieldName>` | `--fi` | (Required) The name of the field to migrate |
```
@alvarosabu if I remember correctly, as migrations are per-component (not per-field) should we set the --field as required?

edodusi replied:
actually `--field` is not an option of the `generate` command

alexjoverm commented on src/commands/migrations/generate/README.md:
```diff
@@ -0,0 +1,166 @@
+.storyblok/
+└── migrations/
+    └── {spaceId}/
+        └── {component-name}-{field}.js
```
@alvarosabu same here... I feel the Readme might be outdated, as the tests and src are designed per-component, while I see here and there in the Readme things are per-field

alvarosabu committed:
docs: update migration generate command documentation

- Removed field-specific references in README
- Updated command usage examples to reflect component-level migration generation
- Simplified file naming conventions in documentation
- Updated manual testing checklist and best practices to align with new approach
- Clarified command options and output structure without field-specific details

edodusi commented on src/commands/migrations/generate/index.ts:
```diff
@@ -0,0 +1,71 @@
+    const { suffix } = options;
+
+    if (!componentName) {
+      handleError(new CommandError(`Please provide the component name as argument --componentName YOUR_COMPONENT_NAME.`), verbose);
```
`--componentName` is not a flag, maybe this is a leftover?

edodusi commented:
Errors like running migrations while not logged in, or giving an incorrect component name, result in the CLI not properly exiting:

```
▲ error No component found with name "simple_components"

ℹ For more information about the error, run the command with the `--verbose` flag

⠸ Generating migration for component simple_components...
```

alvarosabu committed:
fix: improved error handling and spinner failed state

edodusi reviewed (approved):
LGTM
